Extend subgraph_rewriter logic

ansley opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#51226 Extend subgraph_rewriter logic**

- Change output of `replace_pattern` to `List[Match]` reflecting the pattern(s) matched in the original graph
- Ensure that all Callables (not just FunctionType objects) work with the rewriter
- Fix incorrect matching in degenerate case (`test_subgraph_rewriter_correct_output_replacement`)
- Verify that pattern matching works when pattern and original graph are the same

Commit:
Extend subgraph_rewriter logic

[ghstack-poisoned]

Commit:
Update on "Extend subgraph_rewriter logic"




[ghstack-poisoned]

Commit:
Update on "Extend subgraph_rewriter logic"




[ghstack-poisoned]

jamesr66a commented on torch/fx/subgraph_rewriter.py:
```diff
@@ -220,6 +235,9 @@ def overlaps_with_prev_match(match : Match) -> bool:
         return False
 
+    assert matches, ("No matches found between original nn.Module `gm` "
+                     "and pattern `pattern`")
```
I don't think this should be an error, it would break composability, e.g. if you want to run a pipeline of replacements that may or may not appear at each stage, this would break that workflow

jamesr66a replied:
If the user wants to verify that something was matched, we should expose that directly rather than constraining the use cases this can be used for

jamesr66a commented on torch/fx/subgraph_rewriter.py:
```diff
@@ -27,11 +29,15 @@ def matches_subgraph_from_anchor(self, anchor : Node) -> bool:
         self.nodes_map = {}
-        return self._match_nodes(self.pattern_anchor, anchor)
+        return (self._match_nodes(self.pattern_anchor, anchor)
+                and len(self.nodes_map) == len(self.pattern.nodes))
```
What situation is this fixing? It's not immediately apparent to me what this condition ensures

jamesr66a commented on torch/fx/subgraph_rewriter.py:
```diff
@@ -265,15 +283,33 @@ def mark_node_as_replaced(n : Node) -> None:
+            subgraph_output.replace_all_uses_with(copied_output)
+        # CASE 2: The pattern subgraph match extends to the end of the
+        # original graph, so we need to change the current graph's
+        # output Node to reflect the insertion of the replacement graph
```
I don't really understand what's going on here. The comment hints that we're changing something, but it doesn't say what or why, and when I read the code it's not obvious

Commit:
Update on "Extend subgraph_rewriter logic"




[ghstack-poisoned]

jamesr66a commented on torch/fx/subgraph_rewriter.py:
```diff
@@ -55,19 +72,28 @@ def attributes_are_equal(pn : Node, gn : Node) -> bool:
+        if pn.op == "output":
+            # This covers the case in which the original graph Node that
+            # matched to the pattern subgraph's output has more than
+            # one input Node
```
(also discussing offline)

If I understand this correctly, it's basically:

1. We have an assumption that the pattern graph has exactly one output value
2. In the case that we have `pn` is the pattern output node and `gn` is some anchor, we need to check for the possibility of a pattern match on any of the nodes that `gn` consumes.


Are these assumptions correct? In that case I think we should be doing something like:

```
if pn.op == 'output':
  assert len(pn.all_input_nodes) == 1
  pattern_output_val = pn.all_input_nodes[0]
  match_found = any(self._match_nodes(pattern_output_val, gn_) for gn_ in gn.all_input_nodes)
<...>
```

Commit:
Update on "Extend subgraph_rewriter logic"


- Change output of `replace_pattern` to `List[Match]` reflecting the pattern(s) matched in the original graph
- Ensure that all Callables (not just FunctionType objects) work with the rewriter
- Fix incorrect matching in degenerate case (`test_subgraph_rewriter_correct_output_replacement`)
- Verify that pattern matching works when pattern and original graph are the same

[ghstack-poisoned]

jamesr66a commented on torch/fx/subgraph_rewriter.py:
```diff
@@ -60,25 +63,52 @@ def attributes_are_equal(pn : Node, gn : Node) -> bool:
     subgraph (``replacement``).
 
-    Args:
+    Parameters:
```
Shouldn't this be `Args` under the Google python docstring style guide? https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html

jamesr66a commented on torch/fx/subgraph_rewriter.py:
```diff
@@ -60,25 +63,52 @@ def attributes_are_equal(pn : Node, gn : Node) -> bool:
+                match_found = all(self._match_nodes(pn_, gn_) for pn_, gn_
+                                  in zip(pn.all_input_nodes, perm))
+                if match_found:
+                    break
```
We should assert that the pattern `Graph` has exactly one output value and this branch goes away, right?

ansley replied:
I have the `break` in there to prevent us from iterating through the entire list of permutations
